### ISSUE TITLE
Make ":set +c" optional

### ISSUE THREAD

**Issue by vshabanov** [haskell/haskell-mode#1436]
`:set +c` is unconditionally executed at GHCi startup (for `:type-at` in GHC 8+, added in
https://github.com/haskell/haskell-mode/pull/1347) but it considerably slowdowns module loading in large projects.

`Collecting type info for N module(s) ...` takes long time (and GHCi eats a lot of memory). And (although I'm not sure how it's connected) saving of large files becomes painfully slow (I couldn't even move cursor in Emacs for 5-10 seconds).

I've just commented `:set +c` in `haskell-commands.el` and everything went smooth. Could you, please, add an option to disable `:set +c`?

**Comment by gracjan** [comment]
I hate such tradeoffs with a passion. On one hand `:type-at` at friends are mostly useful in projects that you cannot keep in your head, that is larger ones. On the other hand it cannot be used in larger projects due to loading time and memory usage problems.

It basically renders the feature useless in conditions when it is most needed.

**Comment by vshabanov** [comment]
It's a fresh feature that could be optimized in future. Plus it could be useful for novices with small projects.

But at the moment it take about 10 seconds after reloading module. And eats additional 2-4G of RAM after several reloads. I've posted a bug report https://ghc.haskell.org/trac/ghc/ticket/12706 for this. So it worth to add an option till it will be resolved.

**Comment by vshabanov** [comment]
And maybe it worth to add `Collecting type info for N module(s) ...` message, so user could see where time is spent between `Compiling: MyModule` and `OK`. I initially thought that there was a performance bug in haskell-mode, not in GHCi.

**Comment by gracjan** [comment]
Are you able to implement such option? I can review the code if you do.

**Comment by vshabanov** [comment]
Seems that I just need to add `defcustom`. I'll try.

**Comment by vshabanov** [comment]
But adding `Collecting type info for N module(s) ...` message could be too much for me. I have almost no experience with Emacs LISP.

**Comment by fice-t** [comment]
@vshabanov Can you evaluate the following and test to see if the `Collecting type info...` message shows up?

```elisp
(defun haskell-process-live-build (process buffer echo-in-repl)
  "Show live updates for loading files."
  (cond
   ((haskell-process-consume
     process
     (concat "\\[[ ]*\\([0-9]+\\) of \\([0-9]+\\)\\]"
             " Compiling \\([^ ]+\\)[ ]+"
             "( \\([^ ]+\\), \\([^ ]+\\) )[^\r\n]*[\r\n]+"))
    (haskell-process-echo-load-message process buffer echo-in-repl nil)
    t)
   ((haskell-process-consume
     process
     (concat "\\[[ ]*\\([0-9]+\\) of \\([0-9]+\\)\\]"
             " Compiling \\[TH\\] \\([^ ]+\\)[ ]+"
             "( \\([^ ]+\\), \\([^ ]+\\) )[^\r\n]*[\r\n]+"))
    (haskell-process-echo-load-message process buffer echo-in-repl t)
    t)
   ((haskell-process-consume
     process
     "Loading package \\([^ ]+\\) ... linking ... done.\n")
    (haskell-mode-message-line
     (format "Loading: %s"
             (match-string 1 buffer)))
    t)
   ((haskell-process-consume
     process
     "Collecting type info for [[:digit:]]+ module(s) ...")
    (haskell-process-echo-load-message process buffer echo-in-repl nil)
    t)
   ((haskell-process-consume
     process
     "^Preprocessing executables for \\(.+?\\)\\.\\.\\.")
    (let ((msg (format "Preprocessing: %s" (match-string 1 buffer))))
      (haskell-interactive-mode-echo (haskell-process-session process) msg)
      (haskell-mode-message-line msg)))
   ((haskell-process-consume process "Linking \\(.+?\\) \\.\\.\\.")
    (let ((msg (format "Linking: %s" (match-string 1 buffer))))
      (haskell-interactive-mode-echo (haskell-process-session process) msg)
      (haskell-mode-message-line msg)))
   ((haskell-process-consume process "\nBuilding \\(.+?\\)\\.\\.\\.")
    (let ((msg (format "Building: %s" (match-string 1 buffer))))
      (haskell-interactive-mode-echo (haskell-process-session process) msg)
      (haskell-mode-message-line msg)))))
```

**Comment by vshabanov** [comment]
It doesn't show up. I've tried `M-x eval-expression` and `C-M-x` (Evaluate defun) in scratch buffer.

**Comment by fice-t** [comment]
I edited the above according to [this](https://github.com/ghc/ghc/blob/876b00ba25a615423f48b0cf9d443a9fd5dbd6f4/ghc/GHCi/UI/Info.hs#L261). Apparently there's an extra space after the `...`. Please try again with the above.

> It doesn't show up. I've tried M-x eval-expression and C-M-x (Evaluate defun) in scratch buffer.

FYI, you only need to do one of those.

> And I've tried both your scroll branch and master branch of haskell-mode

Either branch is fine -- the changes I made on my `scroll` branch don't interact with this code.

**Comment by vshabanov** [comment]
Now it shows `Compiling: nil (nil)` and then `Unexpected response from haskell process`.

`Collecting type info...` goes after `Ok, modules loaded...` so it breaks `haskell-process-load-complete` by consuming `Ok, ...`.

I've changed
```lisp
   ((haskell-process-consume
     process
     "Collecting type info for [[:digit:]]+ module(s) ... \n")
    (haskell-process-echo-load-message process buffer echo-in-repl nil)
    t)
```
to
```lisp
   ((string-match "\\(Collecting type info for [[:digit:]]+ module(s) ...\\)"
                  (haskell-process-response process)
                  (haskell-process-response-cursor process))
    (haskell-mode-message-line (match-string 1 buffer))
    ;; Do not consume "Ok, modules loaded ... Collecting type info ...",
    ;; just exit.
    nil
    )
```
Not sure whether it's a good solution but it works.

**Comment by fice-t** [comment]
> Now it shows Compiling: nil (nil) and then Unexpected response from haskell process.

Ah, that was an older version of it I had up last night before I saw that message and changed it again. Though the version above may still have the unexpected response.

> Collecting type info... goes after Ok, modules loaded... so it breaks haskell-process-load-complete by consuming Ok, ....

That would change things. I don't have GHC 8 on my machine, so sorry about not knowing that.

> Not sure whether it's a good solution but it works.

The solution might not be optimal, but it seems simple and benign, so if it works for you then feel free to make a PR.

**Comment by vshabanov** [comment]
Added https://github.com/haskell/haskell-mode/pull/1443

**Comment by neongreen** [comment]
I put a $20 bounty on this issue, perhaps it'll motivate someone to fix it if it turns out to be easy: https://www.bountysource.com/issues/38379495-make-set-c-optional

**Comment by gracjan** [comment]
Well, fix is easy: just add another setting. The issue is with overall complexity that grows with every setting added.

As I see the way forward: Lets add this setting, document it and then wait till intero takes over all functionality of haskell-interactive-mode.

**Comment by ivan-m** [comment]
Wait, we're going to let intero take over haskell-interactive-mode?  What about those of us that don't use stack?

**Comment by slpopejoy** [comment]
`:unset +c` is a serviceable workaround whilst this debate rages ...

**Comment by gracjan** [comment]
Fixed by #1515.
